rosbag2 QoS spike findings

emersonknapp opened the pull request:
This document describes our proposed approach to solving the rosbag2 qos story 
Closes #208 
Related to ros2/rosbag2#125

Signed-off-by: Emerson Knapp <emerson.b.knapp@gmail.com>

emersonknapp commented:
Hey @Karsten1987 we're evaluating putting our resources on this to push it through for Foxy, this is our proposed design to solve it - it would be really helpful if you could evaluate it!

piraka9011 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,98 @@
+## Explicit configuration
+
+The above behavior is the most sensible default, but users may always have unanticipated needs.
+To provide maximum flexibility, we also expose a per-topic QoS Profile configuration for each of recording and playback.
```
Do you need to specify as part of the API or CLI?

Karsten1987 replied:
I could envision a cli interface similar to `ros2 topic` where we can have a set of pre-defined profiles such as `--latched <topic1> ... <topicN>` or `--best-effort <topic1> ... <topicN>`. 
Alternatively, I like the idea of specifying a yaml configuration which can potentially hold a more advanced configuration per topic.

emersonknapp replied:
I like the idea of a YAML config better than lots of CLI options, personally

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,98 @@
+The intention of this spike is to propose a solution to https://github.com/ros2/rosbag2/issues/125 - "Use QoS for subscription, recording, and playback"
+
+Currently, rosbag2 attempts to record all topics according to a default QoS Profile.
+This means that it can have a QoS mismatch with publishers (for recording) and subscriptions (for playback) that are not compatible with the default profile.
```
Maybe I am reading this the wrong way, but shouldn't it be the other way around? The subscriptions being used for recording within rosbag are not matching any non-default publishers within your ROS2 system. 
Also, as of now, the publishers of rosbag for playback shouldn't be a problem. Given that every successfully recorded message was recorded with default QoS, playing these messages back with default QoS should be appropriate.

emersonknapp replied:
I think you are misunderstanding how QoS works in DDS (and therefore ROS2)

Publishers in the system Offer some QoS Profile (collection of policies).
The rosbag2 Subscription is currently Requesting the default QoS.
If all Requested default QoS Policies are less strict than all Offered QoS Policies, then a connection will be established with an "actual QoS" of the less-strict Request.
This means that we successfully subscribe to this more-strict Publisher.
Subscriptions in the system that also Request that more-strict QoS Profile will be able to subscribe to the original node.
But, on playback, the rosbag2 Publisher will Offer the less-strict default QoS Profile, and those other Subscriptions (Requesting a more strict Profile), will not be able to subscribe to rosbag2's playback.

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,98 @@
+Using this API we can introspect information about the topics and intelligently handle QoS to not lose information.
+
+The high level design is to use this API at recording time to create the internal subscription with a compatible QoS profile to the published topics, and store the QoS information as topic metadata in the bag.
+At playback time, this metadata can be used to choose an appropriate QoS Profile for the publisher.
```
From your experience - I haven't played around with that new `rmw` feature - are the QoS profiles identical between publishers and subscriptions? That is, the QoS for a subscriber can 'blindly' be recorded and applied to the respective publishers? I could see how that works for most cases (latching or best effort), even though I believe there is a difference between best effort subscription and best effort publishing. 

I think what I want to highlight here is that we might want to think about "choose and appropriate QoS Profile for the publisher" in more detail as this is really the key to it.

emersonknapp replied:
The new API returns a list of all "endpoints" for a topic. This means it provides the exact Offered QoS, from all Publishers, as a list, for a given topic.

The `get_actual_qos` API for a Subscription, on the other hand, may return a value that is not on this Offered list, because the Offering Publishers will negotiate down to match only what we've Requested.

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,98 @@
+We choose to record _all_ of these QoS profiles in the topic metadata for completeness of information.
+This case generalizes the simple case preceding to handle all anticipated situations.
+
+At recording time, we use a utility function to construct a "combined least strictness" QoS Profile that _requests_ the least strict QoS Policy for all discovered publishers, which will cause all publishers to match the recording subscription.
```
Is that possible? A subscription can request a change of QoS settings dynamically for external publishers? That's seems odd to me. Do you have more details on this? Same below in the playback case.

Karsten1987 replied:
So reading this document further (especially the milestone section) I got the impression that the change of QoS is a custom ROS2 thing which should be implemented by each node then?

The problem I see with this is that it's potentially disrupting existing systems. Imagine a navigation system where a camera is used to do localization at a certain QoS and by starting a rosbag record, that QoS is changed. That would imply a change in the localization algorithm as well.

I would advocate for raising an error/warning to the user that multiple QoS settings were detected on the same topic and rosbag doesn't have a clear answer to how to handle this case. Also, in my experience that case of having multiple publishers on the same topic, especially with different QoS, shouldn't occur that often.

emersonknapp replied:
I'm not explicitly talking about a "Change of QoS" anywhere. Requesting a less-strict QoS Profile than the Publisher Offers will get you a less-strict connection, but it won't affect other Subscriptions that Requested the more-strict option. There can be node-to-node connections on the same topic with different "Actual QoS"

emersonknapp replied:
We should be able to successfully subscribe to (record) all publishers to a topic without affecting the behavior of that system.

For the playback case, the proposed solution means that the QoS for a given topic will be merged into the most-strict original Offer. In this case, some nodes on a mixed-QoS topic may receive _more_ messages on that recorded topic than they would have in the original system because we have merged all sources into one and so they successfully match with all publishers when they may have only matched with one originally. A warning about this fact on playback may be appropriate, but I think for 99.99% of cases it should be an acceptable behavior.

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,98 @@
+
+## Deadlines
+
+Assuming rosbag2 is accepted into the `ros_base` variant (https://github.com/ros2/variants/pull/24), we will have the following deadlines:
```
It is merged

emersonknapp replied:
Saw that, I'll update this

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,98 @@
+* Create utility functions to min/max each QoS Policy
+* Create utility function to min/max QoS Profile (wrapper around each Policy function)
+
+## Milestone 2a - QoS serialization / deserialization
```
I consider Milestone 2 more important than Milestone 1 in this case.

emersonknapp replied:
Milestone 1 enables Milestone 2 proper subscription / publisher creation. Could split this out to say that there is actually a separate milestone of just recording the QoS without actively using it. It is separate, but the adaptive playback behavior will depend on it.

emersonknapp commented:
@Karsten1987 sorry, I know it's a long document, but could you give it another read through? I've added background on how QoS Profiles work and rewritten the design proposal to try and use strict terminology. 

Also reordered the milestones to reflect current understanding of importance. I think we should try to hit all of them, but they're in an order that I think is now the best bang-for-buck ordering if we only make it partway through.

mm318 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+
+## Current problem
+
+rosbag2 attempts to record all topics by Requesting the default QoS Profile.
```
Can't the immediate solution be to just record by requesting the "least strict" QoS profile?

To request the "strictest possible" QoS profile, and actually leverage the middleware's QoS features, would be ideal, but a stretch goal for Foxy.

emersonknapp replied:
Yeah, "least strict possible" QoS for recording is probably a decent approach to get immediate relief, however it's got a few side-effects that i'm not sure we want to take on:
* Latching is a requested part of this feature (https://github.com/ros2/rosbag2/issues/129), which this solution won't handle
* We'll never get RELIABLE connections, so we could potentially drop more individual messages than we currently are

mm318 replied:
Okay, and given you're going to implement spawning a "combined least strict" publisher for playback, it's natural to adapt that implementation into spawning a subscription as well.

mm318 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+* rosbag2 Requests VOLATILE (default) Durability on `/topic` (and successfully connects with VOLATILE)
+
+Good, data recorded! But on playback:
+* rosbag2 Offers VOLATILE (default) Durability on `/topic`
```
Similarly, can't the immediate solution be to just playback by offering the "most strict" QoS profile? (Although "most strict" may be a lot harder to define than "least strict".)

mm318 replied:
Okay, I can understand why during playback, you would want to mimic the QoS setup as it was when recorded, as it would be a more truthful replay.

It's not because you want every subscriber listening to the rosbag to receive the messages, it's because you purposely want some subscribers to precisely not receive messages during rosbag replay due to QoS mismatch?

emersonknapp replied:
Yeah, you want to match the original system as closely as possible, within the constraints we have.

emersonknapp replied:
Which means, yeah, that we want to avoid sending data to Subscriptions that wouldn't originally have gotten it, if possible.

mm318 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+* Multiple rosbag2_transport subscriptions, one per QoS Profile
+  * Why? The least strict QoS subscription will receive messages from all publishers, so messages from specific publishers will be duplicated. Messages don't have unique IDs so we can't trivially deduplicate.
+* Choosing only one (not merged/modified) of the QoS profiles to use
+  * Why? This won't necessary catch all topics. QoS Profiles are not monotonically stricter/laxer than each other, only individual Policies within the Profile.
```
Out of the policies that ROS2 supports, wouldn't the "least strict" QoS profile just be:
- History: N/A
- Depth: N/A
- Reliability: BEST_EFFORT
- Durability: VOLATILE
- Liveliness: AUTOMATIC
- Lifespan: N/A
- Deadline: <very large (infinite?) deadline period>

emersonknapp replied:
Default deadline is infinite, so yes that's correct.

And "most strict" offer for playback would look like
* Reliability: RELIABLE
* Durability: TRANSIENT_LOCAL
* Liveliness: MANUAL_BY_TOPIC
* Deadline: this one doesn't seem to have a sensible default, because you can't specify 0 as you'd have to send infinitely fast. Also interesting because you may have recorded multiple publishers, so the playback will have more messages than any original publisher, which means the playback will have a higher frequency

MichaelOrlov replied:
@emersonknapp Challenging question how to properly deduct history depth ?

emersonknapp replied:
History does not affect compatibility so any value can be in the request

MichaelOrlov replied:
@emersonknapp From one hand yes it doesn't affect QoS compatibility but from another hand it could significantly affect results during playback. \
The problem is how to estimate this `any` number.

emersonknapp replied:
I see what you mean - but I think it's outside the scope of this feature since we're focused on compatibility.

MichaelOrlov replied:
I am Ok with it. Just fair warning for whom who will implement it.

emersonknapp commented:
Thanks for the feedback @mm318. I think that the "least/most strict possible" suggestion has merit, but I am not certain if it's what we want to do, I'm on the fence. It would be a very quick way to record more topics, but I don't know if the drawbacks are worse (and I'm not sure how to handle deadline or the liveliness lease duration in the "most strict" case)

MichaelOrlov commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+# Design
+
+`rmw` now has an API to query the Offered QoS Profile for all Publishers to a given topic.
+Using this API we can introspect information about the topics and intelligently handle QoS to avoid both problem cases presented above.
```
Hi @emersonknapp, can you please kindly give an examples of what kind of `rmw` API are you going to use?

emersonknapp replied:
This one https://github.com/ros2/rmw/blob/master/rmw/include/rmw/get_topic_endpoint_info.h#L68

Karsten1987 replied:
I think they're referring to https://github.com/ros2/rmw/pull/186
But +1 for adding these functions to the doc.

MichaelOrlov replied:
@emersonknapp Thank you for the link. As far as I can see you are going to use API from ROS graph. \
It's probably ok for first iteration, but in future it would be nice to have those information from DDS discovery.

MichaelOrlov commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+
+We will use this API to accomplish the following:
+* Record all Offered QoS profiles for a topic into the bag metadata
+* Recording: Process the list of Offered Profiles to create our rosbag2 Subscriptions using a maximally-compatible Request Profile so we successfully subscribe to every Publisher of a topic
```
@emersonknapp It would be very useful if you would also record in metadata QoS settings of rosbag2 subscription which was used during recording.

emersonknapp replied:
do you mean the requested QoS, or the actual QoS, or both?

MichaelOrlov replied:
In document you mentioning that you are going to record all QoS profiles of publishers and derive QoS profile for rosbag2 subscriber. It would be also useful to know with what derived subscriber's QoS settings this data was collected.

emersonknapp replied:
I'm going to resolve this comment, since i'm scoping down the proposal to use the discovered QoS only if all are the same. This chosen one will be recorded.

MichaelOrlov commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+### 1a Serialization/Deserialization
+In `rosbag2_metadata`:
+* Create function to translate `rclcpp::QoS` -> `dict` so that it can be serialized into topic metadata
+* Create function to translate `dict` -> `rclcpp::QoS` so that it can be deserialized at playback time
```
@emersonknapp Can you please clarify, why do you need to serialize/deserialize QoS settings from rclcpp? \
I don't see any rational to do it. \
Since this is metadata for rosbag2 internal usage only and you are no going to send this information over the wires.

emersonknapp replied:
the rosbag2 metadata is written to a YAML file - this requires serialization as it is not a native type

emersonknapp replied:
see https://github.com/ros2/rosbag2/blob/master/rosbag2_storage/src/rosbag2_storage/metadata_io.cpp#L45

Karsten1987 replied:
I guess this is meant to be written into the database in binary form? I could see that it makes sense to store the QoS settings alongside the payload in the storage backend, e.g. sqlite. 
But yes, we don't necessarily have to serialize this QoS in any other form than binary in my opinion. We could also consider putting that information in the `metadata.yaml`, but I tend to having the QoS next to the payload in the storage itself.

MichaelOrlov replied:
@emersonknapp Ohh I see.. \
But this is actually not a serialization/deserialization which I thought in context of DDS. This is just encoding/decoding for YAML file representation it used to be very simple one to one copy to the dedicated field in YAML file.

emersonknapp replied:
@Karsten1987 Is there metadata information in the database about the topics? Are these binary dumps compatible across platforms? I'd like to be careful that a bag recorded on arm32 could be played back on a 64 bit platform. I'm not sure if it works correctly today but I don't want to make any decisions that take us farther away from that.

emersonknapp replied:
encoding as yaml is a serialization operation. I'll be careful to specify "DDS Serialization" when talking about wire protocols.

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+
+At recording time, we use a utility function to construct a "combined least strictness" QoS Profile Request that will be compatible with all the Publishers of a topic.
+
+At playback time, we invert the above logic to construct a "combined greatest strictness" QoS Profile Offer that will be compatible with all the Subscriptions of the original system.
```
How would that work with `latched` topics? If one of the original publishers provided latched data, during playback time all publishers are non-latched? 
I guess that's certainly a plausible solution, given that document here. I am just not sure if that's really user friendly. It might not be transparent for the user to see why their topics get replayed differently (the user might not be the same person who recorded the data).

I am not sure if that's a valid case, but imagine we record latched data and suddenly a second publishers arrives with non-latched data. That would interfere with the already recorded QoS data profile, wouldn't it? Also, even if just triggered by accident, one published message with wrongly set QoS can basically change the complete semantic of the playback.

emersonknapp replied:
I'd rather use the term "transient-local durability" because latched isn't truly a ros2 concept.

For playback, we would Offer transient-local durability, because it is more strict than volatile, so any late-joining subscriptions that want to receive old messages can request transient-local durability, establish a compatible connection, and receive those messages. What this means is that during playback time, if there was one transient-local publisher on a topic originally, then all messages to that topic will be available that way even if they weren't originally.

Yes, multiple publishers can change the semantic of the playback, but only in the direction of more information available, not less.

emersonknapp replied:
Resolving with changed proposal

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+
+In the more difficult case, a topic has multiple Publishers, using two or more distinct QoS Profiles.
+
+At recording time, we use a utility function to construct a "combined least strictness" QoS Profile Request that will be compatible with all the Publishers of a topic.
```
Is this part of the standard? Or how do we determine this "combined least strictness"? Can there be a non-ambiguous solution found?

emersonknapp replied:
Yes, there is an unambiguous solution to this because each Policy within the Profile has an unambiguous compatibility chain

emersonknapp replied:
At least, this is my take on it - I have not seen cases where it's not true

emersonknapp replied:
Resolving since I'm changing the proposal

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+
+## Explicit configuration
+
+The above behavior is a sensible default, but users may always have unanticipated needs.
```
I think my take on this is a more defensive approach, where in the complex case, the data is not recorded and a warning to the user is provided that a clash in QoS is found. The user can then be more explicit about it and specify a suitable QoS via the configuration file.

The rationale: This basically shifts the burden to deal with QoS to the user recording, not replaying. I think the complex case should most likely only very rarely occur in the first place, but if so, I would assume the person recording is better suited to impose an explicit QoS config than the person replaying it, potentially out of context of the live-system. 

but that's just my opinion. There might be good reasons to find a default solution which records data with a suited default QoS. Then I guess though it would make sense to record all discovered QoS to not lose any information.

emersonknapp replied:
I see what you're saying - I wanted to have a more complete out-of-the-box for Foxy, but I could agree with a phased approach where we start with handling the simple case with a warning, then consider automatic behaviors in a second pass.

MichaelOrlov commented:
@emersonknapp Overall good idea :thumbsup: 
I also had an idea that we actually will need to split our recordings in rosbag2 on different streams. Each stream per different QoS settings on publisher side. \
Although to do this proper way we will probably need to change or bring a lot of new API. In this case will need to be able to detect IP address of the publisher and be able to subscribe on topic to dedicated publisher with specific IP.  \
Replay also should be from different streams with separate QoS settings.

Let's do your design first, than we can think about how to organize separate streams for the same topic in rosbag2.

It would be very useful to have statistics at the end of the recording of how many messages did we miss on what topics and if possible from what publishers. \
From user perspective of view such statistic very useful to be able to judge whether recording was successful/acceptable or not.

Karsten1987 commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,163 @@
+
+In `rmw`:
+* Create utility functions to min/max each QoS Policy according to its compatibility scale
+* Create utility function to min/max QoS Profile (min/max each Policy within it)
```
I feel this should be explained in more details. I don't fully understand the scope of this quite yet. Is this a pure utility function and does not effect any existing pub/subs? Can you envision a function declaration of this already?

emersonknapp replied:
Pure utility function that will not affect any other code in ROS 2 - we could put it within rosbag2 if we wanted, but it felt appropriate to make it more widely available. Incomplete pseudocode

```

reliability_qos max_of(reliability_qos a, reliability_qos b) {
  if (a == reliable or b == reliable) {
    return reliable;
  }
  return best_effort;
}

durability_qos max_of(durability_qos a, durability_qos b) {
  if (a == transient_local or b == transient_local) {
    return transient_local;
  }
  return volatile;
}

deadline_qos max_of(deadline_qos a, deadline_qos b) {
  return deadline_qos(std::min(a.as_seconds(), b.as_seconds()));
}

lifespan_qos max_of(lifespan_qos a, lifespan_qos b) {
  return lifespan_qos(std::max(a.as_seconds(), b.as_seconds()));
}


qos_profile maximally_compatible_qos(qos_profile qos_a, qos_profile qos_b) {
  qos_profile out;
  out.history = any_value;
  out.depth = any_value;
  out.reliability = max_of(qos_a.reliability, qos_b.reliability);
  out.durability = max_of(qos_a.durability, qos_b.durabilit);
  

}

qos_profile maximall_compatible_qos(list_of_qos_profiles list) {
  qos_profile max_qos = list[0];
  for (qos_profile next : list[1:]) {
    max_qos = maximally_compatible_qos(max_qos, next);
  }
  return max_qos;
}

```

emersonknapp commented:
Yes, I agree that matching the publisher layout of the original system would be much better, but like you said would require serious changes to the ros2 core whereas this approach is non-intrusive and can be done now.

I also agree that statistics would be useful information, but that should also be a separate improvement with its own design.

emersonknapp commented:
OK, based on the conversation here I am seeing a reasonable case for scoping down to deliver the following:

* record as topic metadata the offered QoS Profile(s) in the system
* automatically handle the "simple case" (all publishers offer same QoS)
* warn the user in the "complex case"  (multiple publishers with different offered QoS)
* allow user to manually specify exact QoS for
  * request on recording
  * offer on playback

Doing just the above should remove all ambiguity, and we can split out discussion of a more complex auto-qos behavior.

emersonknapp commented:
OK, I have updated the proposal to match above comment.

I still believe we want to record all discovered QoS Profiles just so that we have the information.
It appears one unresolved thread remains, and that's - Where should the recorded QoS Profiles be stored, in the database or the metadata yaml?

piraka9011 commented:
> Where should the recorded QoS Profiles be stored, in the database or the metadata yaml?

Given that rosbag2 can support multiple storage plugins, I'm on the side of storing it in the metadata and avoiding any database specific implementation for this (it'll probably increase complexity of the task and for future plugin implementations).

Karsten1987 commented:
> it'll probably increase complexity of the task and for future plugin implementations

That's a fair point, though I would argue that these QoS meta files can potentially become pretty large and cumbersome to read manually anyway. We have the sqlite layout already set as such that we have foreign key relationships between topic and its payload. I can't imagine it to be a big development burden to extend it to also relate to a metadata entry. 

But then again, if we store the QoS in a yaml file, it would also address the CLI problem, where users could simply modify the existing QoS in the same file without specifying a separate "overlay" QoS yaml file.

emersonknapp commented:
> users could simply modify the existing QoS in the same file

Yeah, that'd handle playback. But we do have to still build out all the infrastructure to handle recording anyways, right?

Oh, but for the case of distributing rosbags for others to use, that means you're easily editing and distributing this override information too, in a way that the end user can also easily tweak. That seems like a case in favor of yaml over database?

What data is stored about the topic in the database right now?

piraka9011 commented:
> What data is stored about the topic in the database right now?

You can see it [here](https://github.com/ros2/rosbag2/blob/cf68df2caf27620e6c45f759011d8b163ec3eec8/rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp#L166):

* Name
* Type
* Serialization format

Same stuff in the metadata.

dabonnie commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,160 @@
+
+The intention of this spike is to propose a solution to https://github.com/ros2/rosbag2/issues/125 - "Use QoS for subscription, recording, and playback"
+
+## Necessary background on QoS
```
Very nice section

thomas-moulard commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,160 @@
+## Complex case
+
+In the more difficult case, a topic has multiple Publishers, using two or more distinct QoS Profiles.
+When this happens, fall back to using the default QoS profile to attempt to subscribe, and warn the user that this is happening.
```
Minor comment:

If a new publisher pops up, then `rosbag2` gets notified, query the QoS info, then subscribe. This would lose the data between the initial creation of the topic, and the time `rosbag2` starts recording after resolving the QoS profile. This is already the case in ROS 1, but the additional logic here could make things worse.

emersonknapp replied:
Call out difficulty of handling publishers that join on an already-subscribed topic (with a different QoS than we used to subscribe the first time)

thomas-moulard commented on spike-findings/208-rosbag-qos.md:
````diff
@@ -0,0 +1,160 @@
+  deadline: 100ms
+  liveliness: MANUAL_BY_TOPIC
+  liveliness_lease_duration: 1s
+```
````
using read-only, loaded at startup ROS params would be interesting, so that you can nicely plug that into a launch file (and have one side where to look at config options side by side)

thomas-moulard commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,160 @@
+* Use `rmw_get_publishers_info_by_topic` to get list of all offered QoS profiles for recorded topic
+
+In `rosbag2_metadata`
+* Store list of QoS profiles for the topic as an additional field on topic metadata, alongside name, type, and count
```
Worried about having a canonical representation there. It'd be worth looking at the DDS spec to see if they specify a way to serialize QoS profiles?

emersonknapp replied:
resolving as there is already a conversion layer between rmw_qos_profile and actual dds representation - there's no guarantee we are using DDS, so the rmw struct is the most portable representation for ROS 2

thomas-moulard commented on spike-findings/208-rosbag-qos.md:
```diff
@@ -0,0 +1,160 @@
+
+In `rosbag2_metadata`
+* Store list of QoS profiles for the topic as an additional field on topic metadata, alongside name, type, and count
+* Bump metadata version
```
+ We need to make sure this won't break compat with already existing bag files.

(obvious, but good to call out explicitly)

mm318 replied:
And also forward compatibility with newer versions of ROS2 / `rosbag2`.

emersonknapp commented:
@ros-tooling/approvers  i updated this to make the discovery logic explicit, you think we're good to merge this doc?
